File: C/219-contains-duplicate-ii/main.go
package main

import (
	"fmt"
)

func containsNearbyDuplicate(nums []int, k int) bool {
	if len(nums) == 0 || k <= 0 {
		return false
	}

	buf := make(map[int]int)
	for i := 0; i < k && i < len(nums); i++ {
		_, ok := buf[nums[i]]
		if ok {
			return true
		}
		buf[nums[i]]++
	}
	for i := k; i < len(nums); i++ {
		_, ok := buf[nums[i]]
		if ok {
			return true
		}
		buf[nums[i]]++
		delete(buf, nums[i-k])
	}
	return false
}

func main() {
	fmt.Println("a")
}

File: C/219-contains-duplicate-ii/main.rs
struct Solution {}

use std::collections::HashMap;
impl Solution {
  pub fn contains_nearby_duplicate(nums: Vec<i32>, k: i32) -> bool {
    let mut m: HashMap<i32, usize> = HashMap::new();

    let mut ans: bool = false;
    nums.iter().enumerate().for_each(|(idx, &v)| {
      m.entry(v)
        .and_modify(|x| {
          ans |= (idx - *x) as i32 <= k;
          *x = idx;
        })
        .or_insert(idx);
    });
    ans
  }
}
